### PR TITLE
RemoteExecutionService: conditionally download std{err,out}

### PR DESCRIPTION
**Pull request by sluongng**
When Build without the Bytes is used, only download stderr and stdout
for failed actions.

### CODE REVIEW

**Comment by tjgq** [comment]
An alternative might be to arrange for `--ui_event_filter`/`--output_filter` to prevent the stdout/stderr from being downloaded (i.e., instead of unconditionally raising an event for the stdut/stderr and filtering at reporting time, create a way to ask ahead of time whether the event is of interest). This would keep the options orthogonal, but might be a complex change.

@coeuvre @meisterT thoughts?

File: src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java [hunk 1]
```diff
@@ -1228,13 +1229,15 @@ public InMemoryOutput downloadOutputs(RemoteAction action, RemoteActionResult re
...
     FileOutErr outErr = action.getSpawnExecutionContext().getFileOutErr();
-
-    // Always download the action stdout/stderr.
     FileOutErr tmpOutErr = outErr.childOutErr();
-    List<ListenableFuture<Void>> outErrDownloads =
-        remoteCache.downloadOutErr(context, result.actionResult, tmpOutErr);
-    for (ListenableFuture<Void> future : outErrDownloads) {
-      downloadsBuilder.add(transform(future, (v) -> null, directExecutor()));
+
+    // Skip downloading the action stdout/stderr when remote_download_outputs=minimal and the action is successful.
```
Comment (ivan-golub) [inline src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:1234]: should this apply to `remote_download_outputs=toplevel` too?

File: src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java [hunk 2]
```diff
@@ -1228,13 +1229,15 @@ public InMemoryOutput downloadOutputs(RemoteAction action, RemoteActionResult re
...
+    if (remoteOptions.remoteOutputsMode != RemoteOutputsMode.MINIMAL || result.getExitCode() != 0) {
+      List<ListenableFuture<Void>> outErrDownloads =
+          remoteCache.downloadOutErr(context, result.actionResult, tmpOutErr);
+      for (ListenableFuture<Void> future : outErrDownloads) {
+        downloadsBuilder.add(transform(future, (v) -> null, directExecutor()));
+      }
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:1240]: Even if we don't download them eagerly, we might still want to arrange for these files to be browsable through an output service and/or explicitly downloaded via `--remote_download_regex`.

However, see also my top-level comment - I think we should make a decision on whether gating this by the `RemoteOutputsMode` is the right approach, before we put more effort down this path.

File: src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java (hunk 1 above)
Comment (sluongng) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:1234]: I think that would be a more complicated change that could be addressed in a separate PR. I want to keep this PR simple to discuss the right UX first.

File: src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java (hunk 2 above)
Comment (sluongng) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:1240]: Right now i assume we are targeting toplevel download mode or FUSE for local development UX, where user might have a need to inspect these locally. That left minimal being the ideal for CI/automation setup, where most of these are already available through a BES/RBE service.

**Comment by timothyg-stripe** [comment]
+1 for decoupling this from `--remote_download_outputs` – in theory, someone who runs `--remote_downloads_outputs=minimal` might still want to see the stdout/stderr of their actions (which is the current Bazel behavior).

We are running some tests internally to see if this addresses the performance we've seen with BwoB.

**Comment by sluongng** [comment]
Initially, I thought of using a new flag for this. But going through the existing flag, I found `--remote_download_outputs` to make the most sense for newer users. I mean... it's literally "outputs", and not "artifacts".

Any performance metrics from Stripe setup would be great!

---

Side note: Fresh push to remove some unrelated local dev hunks from the PR.

File: src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java (hunk 2 above)
Comment (sluongng) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:1240]: Giving this a bit more thought: 

- `--ui_event_filters` and `--output_filter` might work, but they are mostly used to control how the outputs are "displayed" and not "downloaded". In some setups, users might want to have separate control knobs for these. (i.e. download the logs but don't display them for some post-run analysis process)

- We can reuse `shouldDownload(result, execPath)` so `--remote_download_regex` is applied as well as covering top_level. Im just not sure if that's the right UX. I will play with it a bit today to see how it goes.

Comment (sluongng) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:1240]: https://github.com/bazelbuild/bazel/commit/3b3a57795ab8292626aeedccc1f4e3f4f494c0b3

Here is a quick draft using the `shouldDownload` approach. I wrote up my findings in the commit's message.

A not-so-obvious thing I missed initially was Bazel does not retain the stdout/stderr files. Instead, it just downloads these into some temp files to dump them out into the terminal and delete them afterward. This means the temp files Path/PathFragment were not eligible to feed into `shouldDownload()` because it uses a different `FileSystem` than the `RemoteOutputChecker`. I managed to piggyback onto the download of output files/dirs to conditionally download the stderr/out, but there are some interesting test failures.

I made the code change in a separate branch because I am not convinced that's the right approach just yet.

I will poke at `--ui_event_filters` approach today. I suspect in a hybrid build where actions are half local, half remote, users might still want to get stderr/out there. But perhaps adding a new value into these flags (i.e. `-remote_stdout,-remote_stderr`) could work. (?)

Comment (sluongng) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:1240]: ah ok, so the shouldDownload approach has issue with `experimental_split_xml_generation` being enabled by default. This checks for stdout existence

https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/exec/StandaloneTestStrategy.java;l=834;drc=300c5867b7d2da1ba32abc20e95662096c2a7a08

and depends on it. Hmm 🤔

Comment (sluongng) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:1240]: Digged a bit more and I found that there are other special interactions for tests when doing none-minimal download mode. For example https://cs.opensource.google/bazel/bazel/+/887e1bd1ad860da91226ba67dfca0eebd8754e8c:src/main/java/com/google/devtools/build/lib/remote/RemoteOutputChecker.java;l=188-191

So I would rather keep this PR at `minimal` mode for now.
